src/main/java/com/arpnetworking/metrics/mad/parsers/IsmClientProtocolRecordParser.java

```diff
@@ -0,0 +1,129 @@
+        } catch (final ConstraintsViolatedException | IllegalArgumentException e) {
+            throw new ParsingException("Could not build record", data.getBody(), e);
+        }
+    }
+
+    private ImmutableMap<String, ? extends Metric> buildMetrics(final ClientV1.Record record) {
+        final ImmutableMap.Builder<String, Metric> metrics = ImmutableMap.builder();
+        processEntries(metrics, record.getCountersList(), MetricType.COUNTER);
+        processEntries(metrics, record.getTimersList(), MetricType.TIMER);
+        processEntries(metrics, record.getGaugesList(), MetricType.GAUGE);
+        return metrics.build();
+    }
+
+    private void processEntries(
+            final ImmutableMap.Builder<String, Metric> metrics,
+            final List<ClientV1.MetricEntry> entries,
+            final MetricType metricType) {
+        for (ClientV1.MetricEntry metricEntry : entries) {
+            final DefaultMetric.Builder metricBuilder = new DefaultMetric.Builder()
+                    .setType(metricType);
+            final List<Quantity> quantities = Lists.newArrayList();
+            for (ClientV1.DoubleQuantity quantity : metricEntry.getSamplesList()) {
+                quantities.add(
+                        new Quantity.Builder()
+                                .setUnit(baseUnit(quantity.getUnit()))
+                                .setValue(quantity.getValue())
+                                .build());
+            }
+
+            metricBuilder.setValues(quantities);
+            metrics.put(metricEntry.getName(), metricBuilder.build());
+        }
+    }
+
+    @Nullable
+    private Unit baseUnit(final ClientV1.CompoundUnit compoundUnit) {
+        if (compoundUnit.getNumeratorCount() > 0) {
+            final ClientV1.Unit selectedUnit = compoundUnit.getNumerator(0);
+            if (selectedUnit.getType() == ClientV1.Unit.Type.UNRECOGNIZED) {
+                return null;
+            }
+            final String unitName;
+            if (selectedUnit.getScale() != ClientV1.Unit.Scale.UNIT && selectedUnit.getScale() != ClientV1.Unit.Scale.UNRECOGNIZED) {
+                unitName = selectedUnit.getScale().name() + selectedUnit.getType().name();
+            } else {
+                unitName = selectedUnit.getType().name();
+            }
+
+            return Unit.valueOf(unitName);
+        }
+        return null;
+    }
+
+    private ImmutableMap<String, String> buildAnnotations(final ClientV1.Record record) {
+        final ImmutableMap.Builder<String, String> annotations = ImmutableMap.builder();
+        for (ClientV1.AnnotationEntry annotationEntry : record.getAnnotationsList()) {
+            annotations.put(annotationEntry.getName(), annotationEntry.getValue());
+        }
+        for (ClientV1.DimensionEntry dimensionEntry : record.getDimensionsList()) {
+            annotations.put(dimensionEntry.getName(), dimensionEntry.getValue());
```

Comment:
Why are we mapping dimensions into annotations? Doesn't the internal model support dimensions separately?